File: backend/app/schemas/training.py
from typing import Optional, Dict, Any
from datetime import datetime
from pydantic import BaseModel
from enum import Enum

class TrainingStatus(str, Enum):
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    STOPPED = "stopped"

class TrainingBase(BaseModel):
    hyperparameters: Optional[Dict[str, Any]] = None
    model_id: int
    dataset_id: int

class TrainingCreate(TrainingBase):
    pass

class TrainingUpdate(BaseModel):
    status: Optional[TrainingStatus] = None
    hyperparameters: Optional[Dict[str, Any]] = None
    error_message: Optional[str] = None
    metrics: Optional[Dict[str, Any]] = None
    cpu_usage: Optional[float] = None
    memory_usage: Optional[float] = None
    gpu_usage: Optional[float] = None

class Training(TrainingBase):
    id: int
    owner_id: int
    project_id: Optional[int] = None
    status: TrainingStatus
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    updated_at: datetime  # Ensuring `updated_at` is always returned
    created_at: datetime
    duration: Optional[float] = None  # Auto-calculated based on start & end times
    epochs_completed: Optional[int] = None
    training_logs: Optional[Dict[str, Any]] = None
    metrics: Optional[Dict[str, Any]] = None
    error_message: Optional[str] = None
    cpu_usage: Optional[float] = None
    memory_usage: Optional[float] = None
    gpu_usage: Optional[float] = None

    class Config:
        from_attributes = True

File: frontend/src/lib/types/training.ts
// src/lib/types/training.ts
export interface TrainingMetrics {
    accuracy: number;
    loss: number;
    precision?: number;
    recall?: number;
    f1_score?: number;
  }
  
  export interface TrainingStatus {
    cpu_usage: number;
    memory_usage: number;
    gpu_usage: number;
    epochs_completed: number;
    current_epoch: number;
    total_epochs: number;
    training_time: number;
    estimated_time_remaining: number;
  }
  
  export interface Training {
    id: number;
    model_id: number;
    dataset_id: number;
    status: TrainingState;
    metrics?: TrainingMetrics;
    hyperparameters: TrainingHyperparameters;
    start_time?: string;
    end_time?: string;
    created_at: string;
    updated_at: string;
    duration?: number;
    error_message?: string;
  }
  
  export type TrainingState = 'queued' | 'running' | 'completed' | 'failed' | 'stopped';
  
  export interface TrainingHyperparameters {
    learning_rate: number;
    batch_size: number;
    epochs: number;
    optimizer: {
      name: string;
      beta1: number;
      beta2: number;
    };
  }
  
  export interface TrainingFormState {
    modelId: string;
    datasetId: string;
    hyperparameters: TrainingHyperparameters;
  }
  
  export interface TrainingCreateRequest {
    model_id: number;
    dataset_id: number;
    hyperparameters: TrainingHyperparameters;
  }